### PR TITLE
Fix support for nullable paths / OsStrings

### PR DESCRIPTION
**Pull request by sdroege**
See https://github.com/gtk-rs/gtk-rs-core/issues/187

This is not complete yet.

### CODE REVIEW

**Comment by sdroege** [comment]
Adding strings into the mix here (`AsRef<str>` instead of `&str`) shows up lots of other strange special cases all over the `gir` code for strings that look like they should be handled more centrally.

Taking a step back here, is it actually worthwhile to have `impl AsRef<Path>` parameters instead of `&Path`? Or `impl AsRef<str>` instead of `&str`?

**Comment by GuillaumeGomez** [comment]
Depends. `AsRef<str>` seems a bit weird to me, however `AsRef<Path>` is pretty nice.

**Comment by sdroege** [comment]
How is it any different? :) Why is one nice and the other not, in which cases does it actually simplify/improve any code? Either avoids an `&` at the caller at the cost of hiding that the function does not actually make use of owned values that are passed in but makes a copy anyway.

**Comment by sdroege** [comment]
I guess because `&str` also impls `AsRef<Path>`, so you can directly pass strings as paths?

**Comment by sdroege** [comment]
Trying to somehow fix this whole mess then :)

**Comment by sdroege** [comment]
This is now ready for review @GuillaumeGomez . PR with the generated changes in https://github.com/gtk-rs/gtk-rs-core/pull/318

**Comment by GuillaumeGomez** [comment]
> I guess because `&str` also impls `AsRef<Path>`, so you can directly pass strings as paths?

It's what I meant: `AsRef<Path>` allows to pass strings for paths, however I don't see a case where `AsRef<str>` can be useful...

File: src/analysis/bounds.rs [hunk 1]
```diff
@@ -203,16 +205,21 @@ impl Bounds {
...
             _ => None,
         }
     }
 
-    fn get_to_glib_extra(bound_type: &BoundType) -> String {
+    fn get_to_glib_extra(
+        bound_type: &BoundType,
+        nullable: bool,
+        instance_parameter: bool,
+    ) -> String {
```
Comment (GuillaumeGomez) [inline src/analysis/bounds.rs:217]: I wonder why we don't return a `&'static str` instead... We don't format anything after all...

**Review by GuillaumeGomez — COMMENTED** [review]
Looks good to me apart from the nit I encountered. Feel free to ignore it to merge this PR but if you do, please open an issue so we can come back to it later on. :)

File: src/analysis/bounds.rs (hunk 1 above)
Comment (sdroege) [reply src/analysis/bounds.rs:217]: For allowing more complicated things later. We had that here at some point in the past.

Changing the type would require changing it in lots of other places too.

**Comment by sdroege** [comment]
Thanks for the review!

**Pull request merged**